### ISSUE TITLE
Convert AccDC to ReactJS

### ISSUE THREAD

**Issue by frastlin** [accdc/accdc#10]
Hello,
I would really love the widgets from AccDC to be ported to a React widget library. This may mean utilizing ES6 more fully.
[Here are the instructions for porting JQuery-like widgets to React](https://reactjs.org/docs/integrating-with-other-libraries.html)
What I'm afraid of is that the whole point of React is to quickly render widgets through the virtual DOM which means that any DOM manipulation should be done through React.
I see two paths forward:

1. Port React using the [integration guide](https://reactjs.org/docs/integrating-with-other-libraries.html) and creating a React Wrapper around AccDC.
2. Change all DOM manipulation to run through React on a version of AccDC.

I've never ported like this, but I desperately want AccDC to be a React framework. That way [companies](https://github.com/facebook/react/wiki/Sites-Using-React) like [Intuit](https://www.intuit.com/) and Facebook can have an accessible UI library to use when their in-house widgets inevitably fail both user testing and accessibility checks.
Currently React is the [fastest adopted](https://www.npmjs.com/npm/state-of-javascript-frameworks-2017-part-1) Javascript framework and it is growing in popularity. React can also be used along with other front-end frameworks such as [Angular](https://blog.rapid7.com/2016/02/03/combining-angularjs-and-reactjs-for-better-applications/) and [Dojo.](http://odoe.net/blog/quick-tip-build-react-dojo-apps/)
Because React is just the "view" in the front-end architecture, it is perfect for AccDC as this is exactly what AccDC is as well.

**Comment by frastlin** [comment]
It looks as if we replaced document in the AccDC API with references to the [virtual DOM](http://reactkungfu.com/2015/10/the-difference-between-virtual-dom-and-dom/), then we can probably just use the AccDC API as it is. I don't know how to get access to the virtual DOM though, so I can look on the React-dom documentation.

Another option is to wrap each component in a react component lifecycle hook (probably componentDidMount) and pass the ref to AccDC to bypass the virtual DOM altogether [as described in this guide.]( https://teamgaslight.com/blog/wrapping-jquery-with-react) (note that this guide is a little out of date, but the principal is the same.)

The other option is just to rewrite the components so they follow the React state architecture (being stateful, using props, and children). This rewriting would probably reduce the code significantly and make it more react-like.
The question is:  
What would best fit the aims of AccDC? Perhaps we can start with trying putting the react components into [lifecycle components](https://reactjs.org/docs/react-component.html) then after that has been done, slowly refactoring the components to be react components.
We could also use the very few accessible widgets that have already been created, such as those by
@davidtheclark
[Tab Panel]( https://github.com/davidtheclark/react-aria-tabpanel)
[Menu Button]( https://github.com/davidtheclark/react-aria-menubutton)
[Modal]( https://github.com/davidtheclark/react-aria-modal)  
@instructure-react
[with their 11 components](https://react.rocks/tag/Accessible)

The power of putting all these components together in a library means users don't need to hunt for what they need, they can just import the components they need and know they will be accessible. None of the existing component libraries have every component accessible and the UX for many of their "accessible" components is terrible. If we could build the beautiful UX of AccDC in React components, it would be amazing.

**Comment by frastlin** [comment]
I have been converting the aria-menu example into React with the lifecycle methods, using the AccDC API:
https://github.com/frastlin/AccDC-React

I have fixed several small things that kept the script from compiling, such as variables being assigned incorrectly or the DOMRect object being read-only.
There is currently a working menu example, but it has no props or children. That will be next.
There are two folders, components and standalone-components. The components are my feeble attempts to create pure react AccDC components. The standalone-components are from the standalone API with the components from the TSG.

**Comment by accdc** [comment]
Awesome!

As an FYI, I haven't forgot, I've been away for about a week on business travel, but I've been looking into the underlying processing as well, and I've created a beta AccDC API update that integrates the rendering of React components directly within AccDC, which makes it easier to create React components that automatically render within React after being processed by AccDC.

So I'm currently creating a bootstrap-react project that converts all of the controles within the bootstrap-jquery project into a fully React controlled project, which I should have completed within a couple of weeks from now. Also included there will be an updated version of the Menus module code as well so that it can tie into this new functionality.

Would you be willing to be a beta tester for this? I would appreciate your knowledge of React when trying it out.

All the best,
Bryan

**Comment by frastlin** [comment]
Of course!
I have somewhat ported the menus and DataGrid to AccDC by wrapping the functions in an onComponentDidMount, but what you are doing sounds more React-like. The API will probably be very similar in the end, but frankly, what I want to do is use the AccDC components in React JSX syntax. For example what I have at:
https://github.com/frastlin/AccDC-React


Menu in App.js:  

~~~
import React, { Component } from 'react';

import Menu from './standalone_components/Menu'
import MenuItem from './standalone_components/MenuItem'

const top_menu = [
<MenuItem primaryText="Food">
	<MenuItem primaryText="Cheese" />
	<MenuItem primaryText="Oranges" />
</MenuItem>,
<MenuItem primaryText="Name" />
]

class App extends Component {
	render() {
		return (
		<div>
		<h1>Menu Example</h1>
			<Menu>
			{top_menu}
			</Menu>
			</div>
		);
	}
}

export default App;
~~~

I've been looking at many different React Frameworks and there is not really any clear syntax for how to do all the different widgets. But a really good thing is to use props.children and the React.Children functionality for widgets.
It would also be nice to pass all the different event handlers into the JSX. For example, in the menu, if there was a onChange and onSelect handler for menus. There are many handlers in the DataGrid, so having those in the syntax would be very helpful as well.

BTW, once I see how you are converting the widgets to React, I can probably do some PR requests as well. I really want this done. I have a couple projects that require menus and DataGrids. I want there to be default CSS for everything, so me as a blind programmer, can show sighted people the widgets without needing the css right away. I'm not 100% sure how all the CSS works, but it would be ideal if we could pass in CSS styling for components both globally and locally. I think the ideal workflow should be:
1. I create the initial app with the default CSS and it looks great.
2. If needed, I have a graphic designer edit the global and local CSS files for my project and make the app look fantastic.

This will speed up my development 1000 times.
I am ready to help, whenever you get the conversion completed. I think the menu and DataGrid are the two most complex widgets, so if you do both of those, I can figure out the others.
I can also write some documentation as well. I would like the process to be as simple as:
1. npm install -g create-react-app
2. create-react-app my-app
3. cd my-app
4. npm install accdc-react --save
5. Write a super simple code, similar to the menu example above.
6. npm start.

That creates a 100% inclusively designed menu!
I'm probably getting a little carried away here, but I think the next step, after we have built the react components, is to create a WYSIWYG editor for either WordPress or copying something like wics or Expo. I had a class last semester where people needed to create a WCAG compliant website right after the introduction to web development, and the nonprogrammers had no idea how to do it. With a WYSIWYG editor, I can get my school to use AccDC which would start creating a core of designers using AccDC.
I would also like to see a widget, similar to:
[This preferences Widget](https://build.fluidproject.org/infusion/demos/uiOptions/)
as part of the AccDC widgets.

**Comment by frastlin** [comment]
Actually from what you said, it sounds as if you are doing what I did in just rendering components through AccDC. I would prefer not to use JQuery as it is just another dependency, but if there is a good reason, I guess it would be OK.
Rendering through the AccDC API is good for now, but frankly, it defeats the whole reason for react in the first place, to use the vertual dom to only render what is really needed. I don't know if it is possible to update the state and have it reflect in the AccDC components.
Here is a good article on exactly
[why people use react](https://medium.freecodecamp.org/yes-react-is-taking-over-front-end-development-the-question-is-why-40837af8ab76)

**Comment by accdc** [comment]
I understand what you mean, and what I'm doing for now is hitting many birds with one stone.

Part of the issue for me to address is all of the module code, since the purpose of AccDC is to normalize accessible widgets across various frameworks equally without having to rewrite all of them just for one library or framework, which makes bug tracking and updates a nightmare to manage since everything has to be tested and built separately.

React is somewhat different, in that they have their own rendering process, and the processing for UI rendering is much faster as a result. Also React is not a widget library as such, and DOM manipulation still has to be done separately.

So, ideally, the most usable solution that doesn't make this project entirely separate from all of the others, is to include the rendering processes within AccDC so that all of the present day modules will automatically be able to use the faster rendering processes when React is available to do so.

The reason being, it isn't going to be feasible for me to rewrite everything that I already have so that it does the same thing in React, but won't work anywhere else, but I can go the other way and have each component that is created within React go through the framework processing within AccDC and actually return rendered React components including all necessary event handlers to ensure accessibility. This will also make it possible for you to add additional event handlers within your React components that increase functionality but have no negative impact on the accessibility handlers added by the AccDC modules.

Also, each time a component is rendered through AccDC as an AccDC Object, you have access to both the standard DOM node of that object via dc.AccDCObj as well as the rendered React component via dc.ReactComponent, and both can be used to process additional rendering such as if you want to perform additional actions on the React object later on after rendering. How this works will become clearer after I finish building it, then I can show examples of these things for you to play with.

Also keep in mind that this is simply to make processing React rendering processes an integral part of the AccDC API, which, when completed, will be available in the Standalone AccDC API version and others, and regardless if jQuery is used or not, will have no impact on the React rendering processes because AccDC will automatically use React for this purpose internally, so you will still have access to all of the benefits of using React.

You will then find it much easier to build custom React components like the menu that you want to build because the processing will automatically be handled internally within AccDC, and then it will be possible to distribute these as you wish. The goal for me at the moment is to enable this functionality internally so that what I've already built can take advantage of this faster processing, while also giving you the ability to create your own widgets that are built entirely using React and still have the same level of functionality.

	Hopefully this makes a bit more sense. It should be clearer when I have this up and running live.

**Comment by frastlin** [comment]
This sounds perfect! So from what I understand is you are rewriting AccDC to either manipulate the dom directly if React is not available and use React if it is.
This means that we get all the advantages of React and AccDC together. This sounds perfect! It will also be useful for applications switching to React from other frameworks.
I really can't wait to use this! Getting the declarative syntax of React with the amazing widgets from AccDC will be perfect for me.

**Comment by frastlin** [comment]
How will this refactoring change the widgets in the coding arena, other than making them faster?
The problem is that I want to make a DataGrid, so currently I copy the DataGrid module from the components folder and place it into a folder with the setup.js. I then convert the module and setup file to export their code. Finally, in the index.js file, I run the setup.js code in the componentDidMount function and in the render function, I have the html that couples with the setup.js file.
Eventually, All the parameters of the setup.js file should be accessible as props to the ~<DataGrid />~ jsx code.
Should I continue doing this, or is there, or will there be a better process?

**Comment by accdc** [comment]
Hi, about the integration of React within AccDC, this is exactly what I am doing. I've already got this running now, I'm just working out the bugs as I convert the syntax for use within a React project using React component objects. Basically it allows bidirectional processing of both React and AccDC objects, both being referenceable from within each other for dual synchronous interaction.

I'm also updating some of the accessible widget modules as needed to enable this functionality, which may be necessary for the grid. I'll need to look at that after I get this part done so you can try the finished beta version for testing. In most cases though, the accessible component widgets will automatically take advantage of this new functionality. For example, since the calendar widget uses a string markup syntax internally, there is a setting within AccDC (dc.RenderUsingReact), which will automatically convert this into a React object for rendering, which is already about 100 times faster than it was before.

I'm also streamlining some of the other widgets to make them easier to use within this environment, and the Menu widget is one I'm going to be visiting soon to see how React component menus can be stacked to automatically be configured as accessible menus. I'll need to experiment with that a bit more later.

At any rate, more news to come later.

**Comment by frastlin** [comment]
How's working out the bugs going? I am really excited to try whatever you have out!
I have two projects I need to complete before the summer ends, one has lots of data grids connected to a data source that users can edit and menu options, similar to spreadsheets, and the second one would use the chat example (or something similar).
I'm holding off doing everything but the back-end currently, because I really want to use AccDC in both.

**Comment by frastlin** [comment]
Are there any code chunks I can test? I would love to start helping in some way.

**Comment by accdc** [comment]
Hi,
I'm sorry for the delay, this is more complicated than I anticipated, and I'm basically rewriting everything to make it conform to more common usage patterns, including all of the component modules so that they will work as integrated parts of React. To give you an idea what I mean, I've pasted my unoficial change log below that is from my notes file as I work on this. Some of these will likely change as I continue working out the details, but you can get an idea what is coming down the pipeline eventually. I'm not sure when specifically since I'm doing all of this in my free time outside of work, so an hour here or there each day is what I'm able to manage at the moment. I'll definitely send a note when it's ready to be tested!



AccDC4X Core Changes

Added the ability to import AccDC directly within the src framework of React so that all dependancies and components can be pulled in as needed including all necessary scripting and accessibility layers without having to expose any external JavaScript within the public folder.


Removed support for $A.html to use $A.insert and $A.insertWithin instead, plus added the ability to pass arrays of DOM nodes to add multiple classes or attributes to many DOM nodes at the same time if desired.

removed block shadow functions do to lack of use.

Refactored the AccDC Object structure to use named regions when boundary is set, and no rendering of offscreen ending tag, plus beginning offscreen node only when forceFocus is true plus role=status and aria-describedby to allow auto announcement by screen readers when rendered and focused.

Added Promise controlled ScriptCache module for dynamic script and stylesheet caching.

moved all local-scoped internal functions into self referencing functions in $A to allow easy tying into external APIs.

added internal $A object method chaining, plus return() method for returning the originally referenced object regardless of type.

added w3C calcNames prototype v2.14 for name computation for dom nodes.

changed isStatic to root as insertion point ref to match common usage patterns.

added $A.on and $.off as aliases of bind and unbind.

updated $A.query to support strings, arrays, objects, and dom nodes and method to loop through each with callback.

Added helper methods for working with interactive ARIA widgets such as idRef methods for adding and removing ID refs within various supported attributes, plus a RovingTabIndex object for handling automatic keyboard support within Menus, Listboxes, Tabs, Radios, Trees, Grids, and Treegrids.

added dc.container as alias of containerDiv

added $A.getParent plus callback to evaluate conditional processing when looking for a parent node with specific properties.

Added various commonly used DOM node handling functions such as $A.insert, $A.insertWithin, $A.append, $A.appendTo, plus others including those specific to handling React components such as $A.mount, $A.mountWithin, $A.unmount, and $A.node for returning the DOM node of any rendered React object, plus the ability to chain these together as needed to change attributes and CSS classes and styles.

added $A.loop function for better processing, handles maps and objects, arrays, single DOM nodes, and strings to iterate through each character with a callback automatically.

Added window.AccDC to be available just like $A as backup within third party APIs.

Changed the $A.reg collection to use a Map() so that any object type can be used as the ID of an AccDC Object, including DOM nodes such as triggering elements.

Added: $A.ReactAccDCObjects, dc.ReactDOMComponent, dc.RenderUsingReact, plus related process handlers for automatically processing React components within AccDC Object constructs when loaded, plus the ability to automatically convert DOM and string markup constructs into React components on the fly.

All modules are being converted into self-contained extendable objects to support dynamic loaders and non-conflicting compartmentalization.

**Comment by frastlin** [comment]
Wow, I understood about half of that, but it looks like a lot of work.
While you are working on these updates, is there a way where I can start using the widgets in my existing apps without needing to change the syntax too much as you push out new updates? I don't mind having the widgets not render in react while I'm still prototyping, but I keep running into times where I would like to use an AccDC component, but adding a basic HTML component is often more simple than building a wrapper around an AccDC component that may change syntax soon.
For example, I started [writing a wrapper that works with React](https://github.com/frastlin/AccDC-React) but I don't know if this is how I should be using AccDC now.
Today I ran into a situation where one of the comboboxes would have been very nice.

**Comment by accdc** [comment]
Hi,
I wish it were that simple, I was trying this week to see if it can be done, but it would cause more problems than not to do this. Basically, as I create these new features, I'm also rewriting the way that the accessible widgets work so that they can be imported directly into react as automatic accessibility layers. This requires a lot of tweaking within the core API so that I can work out the bugs while I add each new widget, so I'm concentrating on all of the widgets within the original AccDC Bootstrap project to start with. When I have these ported over, then the core will be stable enough to test and possibly encorporate into your projects as desired, but until then there are too many unexpected bugs for this to work correctly.

I'm working now to finish the ARIA Menus module, which will make it much easier to build dynamic menu widgets within React using the traditional nested folder structure, and this will tie into all of these new features to make it work. This is coming along well, but it's one of those things that just needs to be finished to work properly.

Now that most of the API changes are made, the time I need now is to get the accessible widgets up and running, which I hope to have time to make progress on this weekend.

More news to come soon though.

**Comment by frastlin** [comment]
Awesome! I'm ready to test when you have something available.

**Comment by frastlin** [comment]
any progress on this? How are the ARIA Menus coming along?

**Comment by accdc** [comment]
Hi,
As a quick update, quite a bit of progress has been made.

The menus are now complete and working great. The module code wasn't the only thing that needed updating though. I've been adding a lot of internal processing to the API, which required me to invent quite a few new features and debug them accordingly, but things appear to be working correctly now.

From the module standpoint, I now have all of the following widgets complete and working properly: Accordion, Banner, Calendar/Datepicker, Menus, Modals, and Tabs.

I'm going to be adding the following soon: Carousel, Footnotes, Popups, Toggles, and Tooltips. Plus if I have time Trees as well.

It actually hasn't been the module rewriting that is taking so much time, but rather the React integration, but now that this is working the module creation should go much faster now that there is a workflow to tie them into.

For example, one of the new features that occurs automatically within the Core API is the automatic addition of bidirectional access. 

This means that you can render a DC object using AccDC that includes a React component for rendering, and this automatically gives you access to the React object from within the lifecycle methods within AccDC, as well as automatic access to the DC object from within the lifecycle methods of your React component. So, within your React component, you can access the DC object directly using this.props.DC from anywhere within the component code, and then you can use this to invoke specific dynamic behaviors on that component such as 
this.props.DC.mount()
this.props.DC.unmount()
this.props.DC.focus()
Plus many others which I'll need to make a list for after I get everything converted. Basically all AccDC API DC object methods and properties can be invoked and set or retrieved from inside your React component using this.props.DC.

I've also completely redesigned the event handling system, and removed all jQuery dependancies including data storage and AJAX.

Now, whenever you specifically bind an event to a DC object from within it's setup properties, it will automatically store a DC object reference on that triggering element, and this will be passed to your event handler as the second parameter. The same thing will occur if you add your own events later on to add functionality to such a triggering element by adding more handlers. When this is done, the event handler will automatically send through the bound DC object connected with that triggering element as the second parameter, even though you didn't actually declare this event during the DC object setup. This makes it possible to add custom handlers to your triggering elements and to control the DC object bound to that element without having to know or retreive the ID for that DC object in advance.

I've also updated all of the internal processes to leverage modern support, as well as related projects that enhance the functionality of AccDC.

For instance, I've added a method, $A.isTouch(), which will automatically detect if the user is presently using a touch screen mobile device such as a phone or a tablet and return either true or false accordingly. This is needed within some of the modules that I'll be working on to handle event handling differences that have to occur in different ways depending on which environment the widget is running within. So to do this, I've imported the resource "current-device", and exposed this within the AccDC object $A.device, which you can then use to access all of the various capabilities documented at
https://www.npmjs.com/package/current-device

I've also added event callbacks within AccDC to expose DC object creation processes, such as
$A.lastCreated // Exposes an array of the last created DC objects
$A.lastCreatedCallback // A function callback that executes once just after the DC creation process completes and pass the $A.lastCreated array to the handler as the first parameter.
dc.onCreated // A handler that can be added to any DC object to return the instantiated DC object just after it is initialized as a standalone object.
$A.lastLoaded // Returns the last DC object that was loaded.

I've also added all of the following methods that prototype into every DC object:
dc.hasClass()
dc.addClass()
dc.remClass()
dc.getAttr()
dc.setAttr()
dc.remAttr()
dc.mount()
dc.unmount()
dc.focus()
dc.widgetType // For setting a control name
dc.autoCloseWidget // For setting this widget to close automatically when another is opened that does not match the same widgetType property name.

There's a lot of other stuff in there too, but this will give you an idea of the big changes. 

I plan to get the remaining modules listed above finished soon, and to have the beta released by the end of August if not sooner.

After the beta is released, I'll continue to work on converting the other modules for use within this environment as well, which I'll push to the same GitHub project when completed. I will also need to add AJAX support after the beta release as well, since I plan to update all of this functionality to utilize the Fetch API for this purpose which will be much faster. 

So the modules that I'll be adding after this beta release are all of the following:
Comboboxes, Data Grids, Listboxes, Radios, Sliders, Inline Form Field Validation, and Progress Bars.

Hopefully this helps explain things a bit. When the beta is ready, I'll upload the project to GitHub and move this issue request to that project for reference, and write some basic documentation to explain how to use these things more fully.

All the best,
Bryan

**Comment by frastlin** [comment]
To say I am eagerly awaiting this is a complete understatement! I'm sure you are going to do this, but can the DC object please have a speak function?
I can't wait to access this!

**Comment by accdc** [comment]
No problem, I'm glad it will be useful!

About a speak function, I hadn't thought of that, but it should be easy to implement. I'll add this as well. :)

**Comment by frastlin** [comment]
I think for now, just use a live region, but I think it would be amazing to eventually have the ability to use either live regions or the web speech API. I built a widget to do this, but it could be made better.
I should start another issue about this, but I would love an area on the website where all the widgets that are officially supported are as well as widgets that are built by the community to work with AccDC can be found in a simple list.
I eventually want to build a [UI Options](https://build.fluidproject.org/infusion/demos/uiOptions/) widget and I would also like to iterate different versions of a speak function people can use.
Eventually I would love to see a CMS built in AccDC. But these should all go into other issues.
Once I can get my hands on the new framework, I've got plans! 👯

**Comment by LaurenceRLewis** [comment]
Now a CMS based on AccDc that I would like to see. 😃

**Comment by frastlin** [comment]
I'm at the point in a little project where a menu is needed. I am either going to build my own menu system, or I would really love to use AccDC if the code is ready.
I'm wondering though, how does one manage moving through menu trees when the next items in the menu tree require a fetch request? For example, I'm building a quick navigation of a map, and I don't want to download all 15000 entries for cities, but I do want to allow people to click on a country, get the regions within that country through a fetch, and when the user selects a region, they get a list of cities in that region.
The current design, and the one in AccDC coding arena, uses an already loaded file or menu tree. How would you build a tree that moves up and down the levels and only calls fetch when required?
I've been trying to figure out a good design for this for a while, and what I'm thinking is that the select callback adds a new menu item to an object of menu objects with a "path-like-this". I have not yet implemented this, but it is what I've been thinking about.

**Comment by accdc** [comment]
[Menu.zip](https://github.com/WhatSock/accdc/files/2295561/Menu.zip)

Hi, the AccDC Menu functionality is already working, so you are welcome to use this for what you want to do. I've attached the component files that show you how they work from a structural level, though I haven't released the API code yet. I am still on track to do this at the end of the month.

I have also already added Fetch API support within AccDC 4X, and you can combine the syntax for loading a fetch request with the dc.setMenu declaration included within it. Below is the syntax for loading three different requests in succession by adding them to the same array, which are automatically stacked.

```
$A.get([

{
url: "/file.html",
data: {returnType: "html", selector: "#tooltip-help"},
success: function(html, promise) {
alert(html);
}
},

{
url: "/file.xml",
data: {returnType: "xml", selector: "group > slide"},
success: function(xml, promise) {
alert(xml);
}
},

{
url: "/file.json",
data: {returnType: "json"},
success: function(json, promise) {
alert(json);
}
}

]);
```

**Comment by frastlin** [comment]
This menu is awesome! Can't wait until the API is out!
So you just put that fetch on the componentDidMount in the Menu component?
It is possible to "Reactify" this component a little more by using children, the spread operator, and whatnot, but they are just abstractions on the current code. I'll probably do it after the first API comes out and you can decide if you like that style or not. Personally, I like to use default settings for most things, so something like:

    <Menu title="Options">
        <MenuItem submenu={ProfileMenu}>Profile</MenuItem>
        <MenuItem>Help</MenuItem>
    </Menu>

But that's my personal preference and extremely easy to do once something like code you uploaded is available.

**Comment by accdc** [comment]
Hi,
I'm closing this issue since the beta has been published at
https://github.com/WhatSock/bootstrap-react

There are some visual styling issues, so if somebody can help me address these I'd really appreciate it. Just do a pull request and I'll be happy to add it! @gericci @frastlin @Decrepidos @dannya @OwenEdwards

**Comment by LaurenceRLewis** [comment]
I’m happy to look at the visual styling.

**Comment by LaurenceRLewis** [comment]
By the way this is a great opportunity for me to learn REACT with AccDc being a library I am very familiar with.

**Comment by accdc** [comment]
Thanks @Decrepidos , I totally spaced on the images, so I've updated all instances where these need a require statement. :)
